### PR TITLE
Add system role support 

### PR DESCRIPTION
**Pull request by dain**
Add support for system level roles in addition to the existing catalog roles.  This updates grammar, adds enabled system roles to identity, and updates file based security to support roles.  Currently roles can only be provided via an extension point in the coordinator, and an SPI will be added in a follow up PR.

### Before
* ROLEs only exist within a catalog
* Catalog only support catalog specific roles
* By default a user runs privileges from ALL roles in all catalogs
* SET ROLE can be used to reduce the enabled roles in a catalog

### After
* ROLEs can be system (global) or within a catalog
* Catalogs that have a catalog access control:
    * Only use catalog specific roles
    * Role management, role grant, table grant, and schema grant commands go to catalog metadata
* Catalogs _without_  a catalog access contro:
    * Only use system roles
    * Role management, role grant, table grant, and schema grant commands go to new system security metadata
* By default a user runs privileges from ALL system roles and ALL roles in all catalogs
* SET ROLE can be used to reduce the active system roles or active roles in a catalog
* Roles can be used in system and catalog level file based access control

### Most Significant Changes
* Catalogs without catalog access control are no longer called for security metadata and instead calls go to the new system security metadata interface
* Require that catalog be explicitly declared in `CREATE ROLE`, `DROP ROLE` and `SET ROLE` with  optional `IN my_catalog` clause for connectors using connector specific roles

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/RevokeRolesTask.java [hunk 1]
```diff
@@ -34,16 +39,25 @@
...
+    public RevokeRolesTask(FeaturesConfig featuresConfig)
+    {
+        legacyCatalogRoles = featuresConfig.isLegacyCatalogRoles();
+    }
+
     @Override
     public String getName()
     {
-        return "GRANT ROLE";
+        return "REVOKE ROLE";
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/execution/RevokeRolesTask.java:58]: Nice. Separate commit?

File: core/trino-main/src/main/java/io/trino/sql/analyzer/FeaturesConfig.java [hunk 2]
```diff
@@ -1088,4 +1090,16 @@ public FeaturesConfig setMergeProjectWithValues(boolean mergeProjectWithValues)
         this.mergeProjectWithValues = mergeProjectWithValues;
         return this;
     }
+
+    public boolean isLegacyCatalogRoles()
+    {
+        return legacyCatalogRoles;
+    }
+
+    @Config("deprecated.legacy-catalog-roles")
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/FeaturesConfig.java:1099]: `@ConfigDescription`

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveRoles.java [hunk 3]
```diff
@@ -13,487 +13,14 @@
...
-import static com.google.common.base.Preconditions.checkState;
-import static io.trino.spi.type.VarcharType.createUnboundedVarcharType;
-import static io.trino.testing.QueryAssertions.assertContains;
-import static io.trino.testing.QueryAssertions.assertEqualsIgnoreOrder;
-import static org.testng.Assert.assertEquals;
-
 @Test(singleThreaded = true)
 public class TestHiveRoles
-        extends AbstractTestQueryFramework
+        extends AbstractTestHiveRoles
```
Comment (kokosing) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveRoles.java:20]: Did you modify any of tests? Is it a simple extraction of abstract class? It is good to have automic changes extracted as separate commit.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHiveRoles.java [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+import static io.trino.spi.type.VarcharType.createUnboundedVarcharType;
+import static io.trino.testing.QueryAssertions.assertContains;
+import static io.trino.testing.QueryAssertions.assertEqualsIgnoreOrder;
+import static org.testng.Assert.assertEquals;
+
+@Test(singleThreaded = true)
+abstract class AbstractTestHiveRoles
+        extends AbstractTestQueryFramework
+{
+    private final boolean legacyCommands;
```
Comment (kokosing) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHiveRoles.java:45]: I would use abstract a method instead. Passing dependencies in constructors usually does not work well in complex cases. So far it is not an issue, but it just a bad habit.

File: client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java [hunk 5]
```diff
@@ -743,17 +743,19 @@ public void testSetRole()
...
-            try (Statement statement = connection.createStatement()) {
-                statement.executeUpdate("SET ROLE bar");
-            }
-            assertEquals(connection.getRoles(), ImmutableMap.of(TEST_CATALOG, new ClientSelectedRole(ClientSelectedRole.Type.ROLE, Optional.of("bar"))));
+
+            // Only hive connector supports roles
+            // try (Statement statement = connection.createStatement()) {
+            //     statement.executeUpdate("SET ROLE bar IN " + TEST_CATALOG);
+            // }
+            // assertEquals(connection.getRoles(), ImmutableMap.of(TEST_CATALOG, new ClientSelectedRole(ClientSelectedRole.Type.ROLE, Optional.of("bar"))));
```
Comment (kokosing) [inline client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java:758]: Do you have a test for that in Hive? If so let's drop it (in separate commit). Or maybe it is already covered with `TestSetRoleTask`?

File: core/trino-main/src/main/java/io/trino/security/AccessControlManager.java [hunk 6]
```diff
@@ -862,112 +862,119 @@ public void checkCanSelectFromColumns(SecurityContext securityContext, Qualified
...
     {
         requireNonNull(securityContext, "securityContext is null");
         requireNonNull(role, "role is null");
         requireNonNull(grantor, "grantor is null");
         requireNonNull(catalogName, "catalogName is null");
 
-        checkCanAccessCatalog(securityContext, catalogName);
-
-        catalogAuthorizationCheck(catalogName, securityContext, (control, context) -> control.checkCanCreateRole(context, role, grantor));
+        if (catalogName.isPresent()) {
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/security/AccessControlManager.java:873]: If catalogName.isEmpty() we should either deny here or throw unsupported.

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java [hunk 7]
```diff
@@ -1759,61 +1765,72 @@ public void dropRole(Session session, String role, Optional<String> catalog)
...
-
         metadata.dropRole(session.toConnectorSession(catalogName), role);
     }
 
     @Override
     public Set<String> listRoles(Session session, Optional<String> catalog)
     {
-        if (catalog.isEmpty()) {
-            return ImmutableSet.of();
+        if (catalog.isPresent()) {
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:1781]: inverse the condition, here all below

File: core/trino-main/src/main/java/io/trino/metadata/MetadataUtil.java [hunk 8]
```diff
@@ -225,6 +226,11 @@ public static void checkRoleExists(Session session, Node node, Metadata metadata
             }
         }
         catalog.ifPresent(catalogName -> getRequiredCatalogHandle(metadata, session, node, catalogName));
+
+        if (catalog.isPresent() && !metadata.isCatalogManagedSecurity(session, catalog.get())) {
+            throw semanticException(NOT_SUPPORTED, node, "Catalog '%s' does not support role management", catalog.get());
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataUtil.java:231]: Can you please provide tests that now statements like `SET ROLE x IN CATALOG c` fails with this error when connector access control does not exists for given catalog?

File: core/trino-main/src/main/java/io/trino/security/AccessControl.java [hunk 9]
```diff
@@ -429,6 +429,7 @@ void checkCanRevokeRoles(SecurityContext context,
 
     /**
      * Check if identity is allowed to set role for specified catalog.
+     * Note: this is not used for system roles.
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/security/AccessControl.java:432]: is this comment still accurate?

File: docs/src/main/sphinx/sql/create-role.rst [hunk 10]
```diff
@@ -9,17 +9,21 @@ Synopsis
...
-``CREATE ROLE`` creates the specified role in the current catalog.
+``CREATE ROLE`` creates the specified role.
 
 The optional ``WITH ADMIN`` clause causes the role to be created with
 the specified user as a role admin. A role admin has permission to drop
 or grant a role. If the optional ``WITH ADMIN`` clause is not
 specified, the role is created with current user as admin.
 
+The optional ``IN catalog`` clause creates the role in a catalog as opposed
+to a system role.
```
Comment (kokosing) [inline docs/src/main/sphinx/sql/create-role.rst:25]: I think we need to explain somewhere what is a difference between catalog role vs system role. Like explain what this actually means for the user and plugin developer.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveRoles.java (hunk 3 above)
Comment (dain) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveRoles.java:20]: I basically changed every line.  I moved the migration to a separate commit, so the commit is still pretty large.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHiveRoles.java (hunk 4 above)
Comment (dain) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHiveRoles.java:45]: In this case I think the constructor with a final boolean field is the right choice.  I understand there are problems if we are talking about more complex objects.

File: client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java (hunk 5 above)
Comment (dain) [reply client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java:758]: There are a bunch of tests for this.  If you are curious search the code base for `"set role`

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java (hunk 7 above)
Comment (dain) [reply core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:1781]: See the comment, this code falls through to the system metadata.  These listing commands are used by information schema, and we want information schema still show roles when the catalog is using system security (it will show system roles instead of nothing).

File: core/trino-main/src/main/java/io/trino/security/AccessControl.java (hunk 9 above)
Comment (dain) [reply core/trino-main/src/main/java/io/trino/security/AccessControl.java:432]: Yes, but it is not relevant so I remove it.  Also, the `@param` below this is wrong.

File: docs/src/main/sphinx/sql/create-role.rst (hunk 10 above)
Comment (dain) [reply docs/src/main/sphinx/sql/create-role.rst:25]: These docs are for end users, so it definitely shouldn't go here.  We don't have any documentation on writing access control implementations.

**Review by kokosing — APPROVED** [review]
The change is massive, so I would encourage you to get another approval too.

File: client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java (hunk 5 above)
Comment (kokosing) [reply client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java]: Then please just drop the code then, instead of commenting.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/FeaturesConfig.java [hunk 11]
```diff
@@ -141,6 +141,8 @@
     private int filterAndProjectMinOutputPageRowCount = 256;
     private int maxGroupingSets = 2048;
 
+    private boolean legacyCatalogRoles;
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/FeaturesConfig.java:144]: I think legacy mode should be default. So far there are no system access control that handle roles (other than file based), so I think we should cause any issues with updates with this.

If you agree then let's rename toggle to `experimental.system-roles` or something like this.

**Review by kokosing — APPROVED** [review]
% comment about disabling the feature by default

File: core/trino-main/src/main/java/io/trino/sql/analyzer/FeaturesConfig.java (hunk 11 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/FeaturesConfig.java:144]: > I think legacy mode should be default.

No, legacy mode should be legacy, and as such, it should be disabled by default. If anyone is relying on the old behavior, they can fall back to the legacy version by setting the toggle. Eventually, we want to remove the legacy form altogether.
